### PR TITLE
Added Join and Leave Buttons for #queue channels

### PR DESCRIPTION
**Pull request by KaoushikMurugan**
Added Join and Leave buttons for queue channel to make it easier to join and leave the queue! Functionality for the /enqueue slash command wasn't changed, there are now two ways to join the queue, either by using the /enqueue command or clicking on a button in the corresponding channel.

https://user-images.githubusercontent.com/63953483/165027003-fb6cb252-c920-4b0d-8524-e6fa1b844cc9.mp4

The "This Interaction Failed" error that shows up after you click the button can be easily fixed with interaction.deferUpdate() (which I have commented out in line 116 in app.ts). But this function causes the bot to crash if the duration between the button click and the call of deferUpdate is more than 3 seconds. I ran the bot locally on my MacBook, so the functions probably took a long time to run because of limited resources (I have had issues with time delay when running a different discord bot on my laptop before). You may try to see if it works when running the bot in the cloud. I shall try to find a permanent fix that doesn't depend on a guaranteed 3 second completion time.

PS. There is some code commented out for stuff I plan on working on soon

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,183 @@
+.. |DiscordLogo| image:: https://img.icons8.com/color/48/000000/discord-logo.png
+   :target: https://discordapp.com
+
+|DiscordLogo| OH-Bot
```
Comment (NoahRoseLedesma) [inline README.rst]: Can you rename this to `BOB`?

File: README.rst [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+======================================
+
+The Discord Office Hour Butler
```
Comment (NoahRoseLedesma) [inline README.rst]: This was the tagline for the old bot, I think `The "Better" Office Hours Bot` would be more appropriate.

File: README.rst [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+their Office Hours through Discord.
+
+We created this bot with the goal of allowing students to effectively
+communicate with their instructors, with the option to be able to ask
+their peers for help while they wait.
+
+Overview
+--------
+
+**OH-Bot** is a server managment automation bot. This means that tasks
```
Comment (NoahRoseLedesma) [inline README.rst]: Use `**BOB**` instead of `**OH-Bot**`

File: README.rst [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+like notifying and moving students when they are ready to been seen are
```
Comment (NoahRoseLedesma) [inline README.rst]: BOB does not move students (the old bot did)

File: README.rst [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+**OH-Bot** is a *self-hotsted* bot - meaning that you will need to host
+and maintain your own instance. See `Quick Start <#quickstart>`__ to
+get started.
+
+The standard OH-Session protocol that we follow: 
+
+#. Instructors Opens OH-Queue 
+#. Students enter queue and enter waiting room 
+#. Instructors dequeue students 
+#. Instructors close queue when OH-Session is over
```
Comment (NoahRoseLedesma) [inline README.rst]: I think these bullet points are outdated. The architecture of BoB is that each instructor is associated with one or more queues, and that doesn't seem to be communicated here.

File: README.rst [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+Waiting Queue System
+--------------------
+
+**OH-Bot** implements a simple *first come first serve* queue where
+student are allowed to enter and leave the queue whenever they like
+while OH is being held.
+
+Server `Template <https://discord.new/kVY9nyrwzV2N>`__
```
Comment (NoahRoseLedesma) [inline README.rst]: This server template should not be used for BOB. Creating a new template is a TODO

File: README.rst [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+There are three roles in the **OH-Bot** server
+
+-  Admin - total control of bot functionality and server interfaces
+-  Instructor - control over OH sessions and locked channels
+-  Student - ability to interface with OH-Queue
+
+Commands
+~~~~~~~~
+
+**OH-Bot** commands have access level based on sender roles
```
Comment (NoahRoseLedesma) [inline README.rst]: These commands are out-of-date

File: README.rst [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+   -  Removes next student from the queue and moves them into the voice
+      channel the sender is currently connected to.
+   -  Removes self from queue if student.
+
+-  ``/help`` - Access Role: [Admin, Instructor]
+
+   -  Sends a DM to the sender containing the above commands and their
+      behavior relative to sender access level
+
+Requirements
```
Comment (NoahRoseLedesma) [inline README.rst:131]: The requirements are out-of-date, the new bot uses node and npm

File: ideas.txt [hunk 9]
```diff
@@ -0,0 +1,52 @@
+options = arguments in discord slash command
```
Comment (NoahRoseLedesma) [inline ideas.txt]: It's good that you have collected ideas. We usually don't track this as code. Create issues on github to track these ideas.

File: src/app.ts [hunk 10]
```diff
@@ -85,8 +88,41 @@ client.on('interactionCreate', async interaction => {
...
-    await ProcessCommand(server, interaction)
+    if(interaction.isCommand()) {
+        await ProcessCommand(server, interaction)
+    }
+    if(interaction.isButton()) {
+        //first 5 characters is the type of interaction, remaining is the name of the queue channel
+        const type = interaction.customId.substring(0,5)
+        const queue_name = interaction.customId.substring(5)
+        //console.log(queue_name + ' app.ts')
+        //console.log(interaction.member?.toString())
```
Comment (NoahRoseLedesma) [inline src/app.ts]: Please remove commented-out code

File: src/app.ts [hunk 11]
```diff
@@ -85,8 +88,41 @@ client.on('interactionCreate', async interaction => {
         server = await JoinGuild(interaction.guild)
     }
     await server.EnsureHasRole(interaction.member as GuildMember)
-
-    await ProcessCommand(server, interaction)
+    if(interaction.isCommand()) {
+        await ProcessCommand(server, interaction)
+    }
+    if(interaction.isButton()) {
+        //first 5 characters is the type of interaction, remaining is the name of the queue channel
```
Comment (NoahRoseLedesma) [inline src/app.ts]: Would it make more sense to have some delimiter character to separate the two pieces of information in the string?

File: src/app.ts [hunk 12]
```diff
@@ -85,8 +88,41 @@ client.on('interactionCreate', async interaction => {
...
+        const type = interaction.customId.substring(0,5)
+        const queue_name = interaction.customId.substring(5)
+        //console.log(queue_name + ' app.ts')
+        //console.log(interaction.member?.toString())
+        if(!(interaction.member instanceof GuildMember)) {
+            await interaction.reply({content: `Erm. Somethings wrong, this shouldn't happen. I'll inform the humaniod that maintains me`, ephemeral: true})
+            console.error(`Recieved an interaction without a member from user ${interaction.user} on server ${interaction.guild}`)
```
Comment (NoahRoseLedesma) [inline src/app.ts:102]: An interaction without a member would mean that the bot received an interaction via direct message which shouldn't be possible. Just FYI.

File: src/app.ts [hunk 13]
```diff
@@ -85,8 +88,41 @@ client.on('interactionCreate', async interaction => {
...
+            return;
+        }
+        //console.log(interaction.type + "in app.ts")
+            var errstr = undefined
+        if(type === 'joinn') {
```
Comment (NoahRoseLedesma) [inline src/app.ts]: Is `joinn` a typo?

File: src/app.ts [hunk 14]
```diff
@@ -85,8 +88,41 @@ client.on('interactionCreate', async interaction => {
...
+            errstr = await server.EnqueueUser(queue_name, interaction.member, interaction.type)
+            //console.log('enqueue function ended')
+        } else if(type === 'leave') {
+            errstr = await server.RemoveMember(queue_name, interaction.member, interaction.type)
+            //console.log('remove function ended')
+        } else {
+            console.log('Received invalid button interaction')
+        }
+        //interaction.deferUpdate()
+        if(errstr !== undefined)
```
Comment (NoahRoseLedesma) [inline src/app.ts]: This seems like a strange way to handle errors. For JS promises we usually have a `.catch` in the promise chain. Also, I wouldn't post the error publicly to the channel.

File: src/queue.ts [hunk 15]
```diff
@@ -34,10 +35,36 @@ export class HelpQueueDisplayManager {
                     messages.clear()
                 }
                 const message_text = this.GetQueueText(queue, queue_members)
+                const buttons = new MessageActionRow()
+                    .addComponents(
+                        new MessageButton()
+                        .setCustomId('joinn' + queue.name)
```
Comment (NoahRoseLedesma) [inline src/queue.ts]: I see what you're doing now. Using a delimiter and splitting would probably be more elegant.

**Review by NoahRoseLedesma — CHANGES_REQUESTED** [review]

File: src/app.ts (hunk 14 above)
Comment (KaoushikMurugan) [reply src/app.ts]: I'm pretty sure there isn't a way to send a message to only the person who pressed the button, since unlike slash commands, there isn't a message to reply to. When I pressed a button without having modified the functions, it caused the bot to crash when the UserError function was called. So the only available options I can think of is either sending the error message in chat or via dms. As for using `.catch`, I'll see if I can use that instead

Comment (NoahRoseLedesma) [reply src/app.ts]: Right, so I would catch the `UserError`  from within this function and handle it before it makes it to the outer exception handler. There isn't a message to reply to, so if you have to communicate maybe shoot a direct message instead.

File: src/app.ts (hunk 11 above)
Comment (KaoushikMurugan) [reply src/app.ts]: used a space as delimiter as category names can not start with a space

File: src/app.ts (hunk 12 above)
Comment (KaoushikMurugan) [reply src/app.ts:102]: Ah ok, I get it now. The reason I had included this code was because it was giving me an error since I hadn't checked if interaction.member is an instance of GuildMember. So I just copied these lines from another part of the code.

File: ideas.txt (hunk 9 above)
Comment (KaoushikMurugan) [reply ideas.txt]: Will do!

File: src/member_state_manager.ts [hunk 16]
```diff
@@ -51,30 +51,24 @@ export class MemberState {
...
             } else if (this.start_helping_timestamp !== null) {
                 throw new UserError('You can\'t join a queue while hosting')
             }
         }
-        //console.log('enqueuetest')
         if(this.current_queue !== null) {
-            //console.log('enqueuetest-case1')
-            return 'Already enqueued'
-            //throw new UserError('Already enqueued')
+            throw 'Already enqueued'
```
Comment (NoahRoseLedesma) [inline src/member_state_manager.ts]: Can we throw a UserError here as well? It seems to me that any of the UserErrors above should apply equally when using the buttons, so I don't see why these errors need to be different.

The difference should be how the error messages are delivered when using buttons vs. slash commands. If you plan on adding more buttons in the future, it might be worthwhile adapting the existing UserError handler to switch where these messages are delivered based on the interaction type.

File: src/queue.ts [hunk 17]
```diff
@@ -38,18 +38,16 @@ export class HelpQueueDisplayManager {
                 const buttons = new MessageActionRow()
                     .addComponents(
                         new MessageButton()
-                        .setCustomId('joinn' + queue.name) 
-                        //intentional misspelling to make the part of the id before queue name to have the same length. 
-                        // didn't shorten into single character like 'j' and 'l' incase we want to add more buttons
-                        //.setEmoji('➕') // U+2795 plus sign emoji
+                        .setCustomId('join ' + queue.name) 
+                        //.setEmoji('➕') // Need to decide on emoji
```
Comment (NoahRoseLedesma) [inline src/queue.ts]: Can you make a choice for this commit? If not maybe remove this line.

File: src/server.ts [hunk 18]
```diff
@@ -82,14 +76,12 @@ export class AttendingServer {
             }
         }
         if (queue === undefined) {
-            return `There is not a queue with the name ${queue_name}`
-            //throw new UserError(`There is not a queue with the name ${queue_name}`)
+            throw `There is not a queue with the name ${queue_name}`
```
Comment (NoahRoseLedesma) [inline src/server.ts]: As per my comment above, this should also probably be a UserError. IIRC throwing strings is generally frowned upon in JS best practices.

File: src/server.ts [hunk 19]
```diff
@@ -305,13 +236,13 @@ private async UpdateNotifQueue(member: GuildMember, queue_name: number): Promise
             }
         }
         if (queue === undefined) {
-            return `There is not a queue with the name ${queue_name}`
+            throw `There is not a queue with the name ${queue_name}`
```
Comment (NoahRoseLedesma) [inline src/server.ts]: Throw UserError

File: .gitignore [hunk 20]
```diff
@@ -4,4 +4,7 @@ node_modules/
 .env
 .nyc_output/
 .vscode/
-gcs_service_account_key.json
\ No newline at end of file
+gcs_service_account_key.json
+botkey.txt
```
Comment (NoahRoseLedesma) [inline .gitignore]: nit: let's not include files you use in your local development environment (botkey.txt, ideas.txt, src/server_test.ts) in the gitignore. Ideally, the gitignore should only contain files any developer working on this project would have.

File: ideas.txt (hunk 9 above)
Comment (NoahRoseLedesma) [inline ideas.txt]: It doesn't look like this file actually got deleted from the commit. IIRC you need to run `git rm ideas.txt`

File: src/app.ts [hunk 21]
```diff
@@ -85,8 +88,38 @@ client.on('interactionCreate', async interaction => {
...
+        } else if(type === 'leave') {
+             await server.RemoveMember(queue_name, interaction.member, interaction.type).catch((errstr : Error) => {
+                if(interaction.member instanceof GuildMember && errstr.name == 'UserError') {
+                    interaction.member.send(errstr.message)
+                }
+             })
+        } else {
+            console.error('Received invalid button interaction')
+        }
+        //interaction.deferUpdate()
```
Comment (NoahRoseLedesma) [inline src/app.ts]: nit: Remove this comment

File: src/server.ts [hunk 22]
```diff
@@ -13,6 +13,7 @@ export class AttendingServer {
     private attendance_doc: GoogleSpreadsheet | null
     private attendance_sheet: GoogleSpreadsheetWorksheet | null = null
 
+
```
Comment (NoahRoseLedesma) [inline src/server.ts]: nit: remove this blank line

File: src/server.ts [hunk 23]
```diff
@@ -53,17 +54,36 @@ export class AttendingServer {
...
             if (queue.Has(member)) {
                 queue_count++
-                return queue.Remove(member)
+                return queue.Remove(member, interaction_type)
             }
         }))
         return queue_count
     }
 
+    async RemoveMember(queue_name: string, member: GuildMember, interaction_type: string): Promise<void> {
```
Comment (NoahRoseLedesma) [inline src/server.ts]: Why is this necessary? The `RemoveMemberFromQueues` function already exists. It is what is used for `/leave`. It seems like the slash command and the button should have the same behavior.

File: src/server.ts [hunk 24]
```diff
@@ -53,17 +54,36 @@ export class AttendingServer {
...
+        const queue = this.queues.find(queue => queue.name == queue_name)
+        if(interaction_type === 'APPLICATION_COMMAND'){
```
Comment (NoahRoseLedesma) [inline src/server.ts]: The conditions here seem redundant. Is checking the interaction type necessary?

File: src/server.ts [hunk 25]
```diff
@@ -53,17 +54,36 @@ export class AttendingServer {
...
+            if (queue === undefined) {
+                throw new UserError(`There is not a queue with the name ${queue_name}`)
+            }
+            if (!queue.is_open) {
+                throw new UserError(`The queue "${queue_name}" is currently closed.`)
```
Comment (NoahRoseLedesma) [inline src/server.ts]: Students can still be on a queue while it is closed. This actually happens fairly frequently in Matt's classes. We want them to be able to keep their spot in line in-between office hours. Similarly, I see no reason why a student can't leave the queue if its closed.

**Review by NoahRoseLedesma — CHANGES_REQUESTED** [review]

File: src/app.ts (hunk 21 above)
Comment (KaoushikMurugan) [reply src/app.ts]: this comment was left on purpose. it currently causes the bot to crash if it takes too long to do a enqueue or leave when I run the bot locally (my laptop is pretty slow when running bots). This line of code should remove the "this interaction failed" error that shows up after you press the button.

File: src/app.ts [hunk 26]
```diff
@@ -85,8 +88,38 @@ client.on('interactionCreate', async interaction => {
         server = await JoinGuild(interaction.guild)
     }
     await server.EnsureHasRole(interaction.member as GuildMember)
+    if(interaction.isCommand()) {
+        await ProcessCommand(server, interaction)
+    }
+    if(interaction.isButton()) {
```
Comment (NoahRoseLedesma) [inline src/app.ts]: This should probably be an `else if`.

File: src/app.ts (hunk 21 above)
Comment (NoahRoseLedesma) [reply src/app.ts]: Deferring the reply makes sense, that's what happens in the `ProcessCommand` path. It would make sense to defer the button interaction update as well. Lets uncomment this and put it as the first line after `if(type == 'join')`

File: src/server.ts (hunk 24 above)
Comment (KaoushikMurugan) [reply src/server.ts]: removed the checking on interaction type since now we're returning UserError regardless of it's a slash command or a button press

File: src/server.ts (hunk 25 above)
Comment (KaoushikMurugan) [reply src/server.ts]: remove those error cases

File: src/server.ts (hunk 23 above)
Comment (KaoushikMurugan) [reply src/server.ts]: the /leave command when called where-ever on the server will remove the person from the queue. But there's a leave queue button for each queue. So I don't want a user to press the leave button for one queue and get kicked out of another. Also I'm planning on possibly allowing someone to join more than one queue so it might be useful later

**Comment by NoahRoseLedesma** [comment]
For now seeing as how students can only join a single queue, let’s stick to
using the same method.

If in the future you want to allow students to be in multiple queues, that
can be worked out. I think it’s tangential to this PR.

File: src/queue.ts [hunk 27]
```diff
@@ -19,8 +19,8 @@ export class HelpQueueDisplayManager {
         const status_line = `The queue is **${queue.is_open ? 'OPEN' : 'CLOSED'}**. There ${quant_prase} in the queue.\n`
         if(queue.length > 0) {
             const table = new AsciiTable()
-            table.setHeading('Position', 'Name')
-            queue_members.forEach((state, idx) => table.addRow(idx + 1, state.member.user.username))
+            table.setHeading('Position', 'Username', 'Nickname')
+            queue_members.forEach((state, idx) => table.addRow(idx + 1, state.member.user.tag, state.member.user.username))
```
Comment (NoahRoseLedesma) [inline src/queue.ts]: I think you're looking for [displayName](https://discord.js.org/#/docs/discord.js/stable/class/GuildMember?scrollTo=displayName). I think it's redundant to have a username and nickname. Maybe just use displayName.

If you want to include this in a separate PR that's ok too. Technically it should be but I won't complain if you make that change here.

**Review by NoahRoseLedesma — CHANGES_REQUESTED** [review]
Please remove `.DS_Store` and `src/.DS_Store`. Adding `.DS_Store` to the gitignore would be good idea.

File: src/queue.ts (hunk 27 above)
Comment (KaoushikMurugan) [reply src/queue.ts]: I think I did something wrong when reverting, since I hadn't meant for this to be in this pull request. I'll remove it

File: .gitignore [hunk 28]
```diff
@@ -6,3 +6,4 @@ node_modules/
 .vscode/
 gcs_service_account_key.json
 .DS_Store
+src/.DS_Store
```
Comment (NoahRoseLedesma) [inline .gitignore]: You can just put `.DS_Store`, (i.e. what you did on line 8) that will ignore that file in any directory.
